src/plugins/dist_matrixops/dist_diag.cpp

```diff
@@ -0,0 +1,576 @@
+#include <phylanx/plugins/dist_matrixops/tile_calculation_helper.hpp>
+#include <phylanx/util/detail/range_dimension.hpp>
+#include <phylanx/util/distributed_matrix.hpp>
+#include <phylanx/util/distributed_vector.hpp>
+
+#include <hpx/assertion.hpp>
+#include <hpx/errors/throw_exception.hpp>
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace dist_matrixops { namespace primitives {
+    ///////////////////////////////////////////////////////////////////////////
+    execution_tree::match_pattern_type const dist_diag::match_data = 
+    {
+        hpx::util::make_tuple("diag_d", std::vector<std::string>{R"(
+                diag_d(
+                    _1_arr,
+                    __arg(_2_k, 0),
+                    __arg(_3_tiling_type, "sym"),
+                    __arg(_4_tile_index, find_here()),
+                    __arg(_5_numtiles, num_localities())
+                )
+            )"},
+            &create_dist_diag,
+            &execution_tree::create_primitive<dist_diag>, R"(
+            arr, k, tiling_type, tile_index, numtiles
+            Args:
+                arr (array): a distributed array. A vector or a matrix.
+                k (int, optional): The default is 0. Denote the diagonal above
+                    the main diagonal when k > 0 and the diagonal below the main
+                    diagonal when k < 0.
+                tiling_type (string, optional): defaults to `sym`. Other
+                    options are `row` or `column` tiling.
+                tile_index (int, optional): the tile index we need to generate
+                    the diag array for. A non-negative integer. If not
+                    given, it sets to current locality.
+                numtiles (int, optional): number of tiles of the returned array.
+                    If not given it sets to the number of localities in the
+                    application.
+            Returns:
+
+            A 1-D array of its k-th diagonal when a is a 2-D array; a 2-D array
+            with a on the k-th diagonal.)")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    dist_diag::dist_diag(
+        execution_tree::primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+
+    namespace detail
+    {
+        //static std::atomic<std::size_t> diag_count(0);
```

Comment:
I think it would be best if you put these functions in an hpp file, so both retile and diag access it.

Reply:
Already put them in a new hpp file, called idx_tile_calculation_helper.hpp